crates/bifrost/src/providers/replicated_loglet/tasks/trim.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use tokio::task::JoinSet;
+use tracing::{debug, trace, warn};
+
+use restate_core::network::{Incoming, Networking, TransportConnect};
+use restate_core::task_center;
+use restate_types::config::Configuration;
+use restate_types::logs::{LogletOffset, SequenceNumber};
+use restate_types::net::log_server::{LogServerRequestHeader, Status, Trim, Trimmed};
+use restate_types::replicated_loglet::{EffectiveNodeSet, NodeSet, ReplicatedLogletParams};
+
+use crate::loglet::util::TailOffsetWatch;
+use crate::loglet::OperationError;
+use crate::providers::replicated_loglet::replication::NodeSetChecker;
+use crate::providers::replicated_loglet::rpc_routers::LogServersRpc;
+use crate::providers::replicated_loglet::tasks::util::RunOnSingleNode;
+
+use super::util::Disposition;
+
+#[derive(Debug, thiserror::Error)]
+#[error("trim loglet operation failed")]
+struct TrimError;
+
+/// Sends a trim request to as many log-servers in the nodeset.
+///
+/// We broadcast the trim to all nodes that we can, but only wait for write-quorum
+/// responses before acknowleding the trim.
+///
+/// The trim operation is idempotent. It's safe to trim a loglet if it's already partially or fully
+/// trimmed and if it's sealed.
+///
+/// Note that the trim task will clip the input trim-point to the last known global-tail of the
+/// loglet, it'll not attempt to perform a FindTail internally, the actual trim point
+/// might be higher if a node in the nodeset has higher trim point and it wasn't visited during
+/// this trim write-quorum check.
+///
+/// In other words, the trim operation confirms that at least write-quorum nodes in the nodeset
+/// are trimmed to the requested (possibly clamped) trim point or higher.
+///
+/// Any call to `get_trim_point()` that happen after this task should return the clamped trim point
+/// or higher.
```

Comment:
Is this really guaranteed? If I understood `GetTrimPointTask` correctly, then we try to reach write-quorum or f-majority but if we don't obtain it and have received at least one response, then we would return this value. So can't it happen that we reach write-quorum here and then when calling `get_trim_point` we only get responses from nodes that weren't part of this write quorum?

Reply:
Correct. I'll update the comment to reflect that it's best-effort